clustermesh: split the generic logic from the specific part

giorio94 opened the pull request:
This PR refactors the clustermesh subsystem, splitting it into a generic part (`pkg/clustermesh`), which is responsible for watching the configuration directory and establishing the connections to remote etcd endpoints, and a clustermesh specific part (`pkg/clustermesh/clustermesh`), which handles the actual synchronization of the required information). This is a preparatory change to allow reusing the generic part for the upcoming kvstoremesh implementation, reducing code duplication and maintenance burden.

giorio94 committed:
clustermesh: split the generic logic from the specific part

This commit refactors the clustermesh subsystem, splitting it into a
generic part (pkg/clustermesh/internal), which is responsible for watching
the configuration directory and establishing the connections to remote
etcd endpoints, and a clustermesh specific part (pkg/clustermesh),
which handles the actual synchronization of the required information).
Additionally, the helper methods dealing with the setting and retrieval
of the CiliumClusterConfig are moved to a separate `clustermesh/utils`
package. This is a preparatory change to allow reusing the generic part
for the upcoming kvstoremesh implementation, reducing code duplication and
maintenance burden.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

joamaki commented on pkg/clustermesh/clustermesh/clustermesh.go:
```diff
@@ -0,0 +1,207 @@
+
+const subsystem = "clustermesh"
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, subsystem)
```
`cell.Module` provides `logrus.FieldLogger` with subsys scoped to module id. Using it will make sure logs match module identifiers and it'll make it easier in the future to use potentially different or augmented logger etc. when it's injected. But since we're mostly moving things around here I don't think it makes sense to change that in the same PR, but perhaps something to keep in mind.

giorio94 replied:
Makes definitely sense to me. I would leave this as a followup (as you also mentioned) to avoid increasing the scope of this PR.

joamaki commented on pkg/clustermesh/clustermesh/clustermesh.go:
```diff
@@ -0,0 +1,207 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package clustermesh
```
`pkg/clustermesh/clustermesh/clustermesh.go` isn't great. Would it be possible to make this `internal` package with public APIs it implements exposed from `pkg/clustermesh` as interface types for example? Or wait, is it the other way around?

giorio94 replied:
Yeah, I also personally find `pkg/clustermesh/clustermesh` to be not that great. I'm not sure about a better one though. One possibility could be to just keep everything in the base clustermesh package, losing a bit of separation (and have the kvstoremesh stuff in `pkg/clustermesh/kvstoremesh`). Otherwise I could do the other way round and move the generic bits in `pkg/clustermesh/internal` or `pkg/clustermesh/generic` (is this what you were proposing?). WDYT?

joamaki replied:
`internal` or `shared` or `generic` package seems like it could work nicely. Hard to say what's best though, so I'll leave it to you.

giorio94 replied:
I've moved the generic part to the `pkg/clustermesh/internal` package, and restored back the business logic in `pkg/clustermesh`, adapting variables names accordingly. I've additionally extracted a few clusterconfig-related utilities in a `pkg/clustermesh/utils` package, as they are also used by the clustermesh-apiserver.

YutaroHayakawa commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -119,113 +112,30 @@ func IsClusterConfigRequired(ctx context.Context, backend kvstore.BackendOperati
 
-	if c.ClusterMeshConfig == "" {
+func (cm *GenericClusterMesh) Start(hive.HookContext) error {
+	if cm.conf.ClusterMeshConfig == "" {
```
What's the situation that `ClusterMeshConfig == ""`? Can't we move this check to `NewGenericClusterMesh`? Same for `Stop`.

giorio94 replied:
Essentially that happens when the flag for the path to the clustermesh config is unset. In that case, the clustermesh subsystem should not be started. I've moved back the check directly to `NewClusterMesh` to restore the pre-existing behavior.

YutaroHayakawa commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -356,64 +246,15 @@ func (cm *ClusterMesh) NumReadyClusters() int {
 }
 
-func (cm *ClusterMesh) canConnect(name string, config *cmtypes.CiliumClusterConfig) error {
+func (cm *GenericClusterMesh) Iterator(fn func(RemoteCluster) error) error {
```
Nit: For me, `Iterator` is the name of the function that returns an iterator object. I'd name this kind of function with `ForEachRemoteCluster` or something. Just a matter of the preference though 😓

giorio94 replied:
Yeah, sounds good to me. I've renamed this method to `ForEachRemoteCluster`.

YutaroHayakawa reviewed (approved):
LGTM with nits 👍

nebril commented on pkg/clustermesh/internal/remote_cluster.go:
```diff
@@ -0,0 +1,377 @@
+	// condition.
+	go func() {
+		if backend != nil {
+			backend.Close(context.TODO())
```
Nit: could we use `context.Background()` here? It seems to be appropriate for async resource release.

nebril commented on pkg/clustermesh/internal/remote_cluster.go:
```diff
@@ -0,0 +1,377 @@
+		for {
+			select {
+			// terminate routine when remote cluster is removed
+			case _, ok := <-rc.changed:
```
Please correct me if I am wrong, but I think that this channel receive will race with the receive from line 273. If this goroutine happens to get in `select` while the goroutine above is not executing and there is a message in `rc.changed` channel buffer, it will get the message from channel, check `ok` value to be `true` (since the channel is not closed) and this goroutine will return while `restartRemoteConnection` would not be called above.

To fix that behaviour, I would suggest merging both goroutines into one so that there is only one select.

giorio94 replied:
Yes, you are right. It seems to me that this logic is also possibly prone to other race conditions, as I've seen connections being restarted multiple times in a row following a single failure. I'll refactor it make it more robust.

giorio94 replied:
~~9668440e8b98e32cd4a5281e8cd25e689a07bafc~~ faebb6f5368610a78c56bd59626dc37e088cc864 addresses this issue refactoring the watchdog logic. I've still maintained the existing `changed` channel logic when an update is detected to avoid introducing too many changes in this PR which is already large due to moving things around.

giorio94 replied:
🤔 There's an issue with the new logic. Let me fix it

giorio94 replied:
Error cases should be correctly handled now.

nebril reviewed (changes requested):
Looks good overall, I think I might have found one bug around restarting remote connections.

giorio94 committed:
clustermesh: fix spurious kvstore connection restarts on errors

The cilium agent implements a watchdog mechanism to restart the
connection to remote kvstores in case of connectivity errors. Yet, the
current implementation is prone to a possible race conditions causing
spurious restarts. Specifically, when an error is detected, it triggers
the restart of the controller responsible for connecting to the kvstore,
and then iterates again, supposedly waiting until the connection is
established (i.e., `rc.backend` is no longer nil). Still, `rc.backend`
is asynchronously reset to nil when the controller is executed, causing
the new iteration of the watchdog to possibly still use the errors
channel of the old connection. This commit fixes this issue moving the
watchdog logic to a separate function which is started by the
controller, and terminates when the corresponding context is canceled
(i.e., the controller is restarted). This also removes the race
condition between the two goroutines reading from the `rc.changed`
channel.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

YutaroHayakawa reviewed (approved):
🚀

nebril reviewed (approved):
LGTM!

joamaki reviewed (approved):
Structure looks good!
